splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java

```diff
@@ -29,108 +29,128 @@
                 _maxReservedTimestamp = timestampBlockManager.initialize();
-				_timestampCounter.set(_maxReservedTimestamp + TIMESTAMP_INCREMENT);
-			}
-			try {
-				registerJMX();
-			} catch (Exception e) {
+                _timestampCounter.updateAndGet(oldTS -> Math.max(oldTS, _maxReservedTimestamp + TIMESTAMP_INCREMENT));
+            }
+            try {
+                registerJMX();
+            } catch (Exception e) {
                 SpliceLogUtils.error(LOG, "Unable to register Timestamp Generator with JMX. Service will function but metrics will not be available.");
-			}
+            }
+    }
+
+    public long getCurrentTimestamp() {
+        return _timestampCounter.get();
+    }
+
+    private static long roundUp(long timestamp) {
+        assert timestamp >= 0;
+        long remainder = timestamp % TIMESTAMP_INCREMENT;
+        if (remainder == 0) {
+            return timestamp;
+        }
+        return timestamp - remainder + TIMESTAMP_INCREMENT;
+    }
+
+    public synchronized void bumpTimestamp(long newTimestamp) throws TimestampIOException {
+        newTimestamp = roundUp(newTimestamp);
+        if (newTimestamp > _timestampCounter.get()) {
+            if (newTimestamp> _maxReservedTimestamp) {
+                reserveNextBlock(newTimestamp);
+            }
+            long currentTimestampCounter;
+            while ((currentTimestampCounter = _timestampCounter.get()) < newTimestamp) {
+                _timestampCounter.compareAndSet(currentTimestampCounter, newTimestamp);
+            }
+        }
     }
 
-	public long getNextTimestamp(boolean refresh, boolean increment) throws TimestampIOException {
-		if (refresh) {
-            _maxReservedTimestamp = timestampBlockManager.initialize();
-			_timestampCounter.set(_maxReservedTimestamp);
-		}
-
-		long nextTS = increment ? _timestampCounter.addAndGet(TIMESTAMP_INCREMENT): _timestampCounter.get();
-		long maxTS = _maxReservedTimestamp; // avoid the double volatile read
-		if (nextTS > maxTS) {
-			reserveNextBlock(maxTS);
-		}
-		_numTimestampsCreated.incrementAndGet(); // JMX metric
-		return nextTS;
-	}
-
-    private void reserveNextBlock(long priorMaxReservedTimestamp) throws TimestampIOException {
-        synchronized(this)  {
-            if (_maxReservedTimestamp > priorMaxReservedTimestamp) return; // some other thread got there first
-            long nextMax = _maxReservedTimestamp + blockSize;
-            timestampBlockManager.reserveNextBlock(nextMax);
+    public long getNextTimestamp() throws TimestampIOException {
+        long nextTS = _timestampCounter.addAndGet(TIMESTAMP_INCREMENT);
+        if (nextTS > _maxReservedTimestamp) {
+            reserveNextBlock(nextTS);
+        }
+        _numTimestampsCreated.incrementAndGet(); // JMX metric
+        return nextTS;
+    }
+
+    private synchronized void reserveNextBlock(long nextTS) throws TimestampIOException {
+        while (nextTS > _maxReservedTimestamp) {
```

Comment:
"while" is redundant here, "if" would suffice

Reply:
done